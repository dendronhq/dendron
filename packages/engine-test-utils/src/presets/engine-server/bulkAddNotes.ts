import { NoteUtils } from "@dendronhq/common-all";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import _ from "lodash";

const SCHEMAS = {};
const NOTES = {
  BASIC: new TestPresetEntryV4(async ({ vaults, engine }) => {
    const vault = vaults[0];
    const orig = _.size(engine.notes);
    const note1 = NoteUtils.create({
      id: "bar1",
      fname: "bar1",
      created: 1,
      updated: 1,
      vault,
    });
    const note2 = NoteUtils.create({
      id: "bar2",
      fname: "bar2",
      created: 1,
      updated: 1,
      vault,
    });
    await engine.bulkAddNotes({ notes: [note1, note2] });
    return [
      {
        actual: _.size(engine.notes),
        expected: orig + 2,
        msg: "should be 2 more notes",
      },
      {
        expected: engine.notes["bar1"],
        actual: note1,
      },
    ];
  }),
};
export const ENGINE_BULK_ADD_NOTES_PRESETS = {
  NOTES,
  SCHEMAS,
};
