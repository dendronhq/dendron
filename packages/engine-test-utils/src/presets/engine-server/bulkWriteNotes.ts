import { NoteUtils } from "@dendronhq/common-all";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import _ from "lodash";

const SCHEMAS = {};
const NOTES = {
  BASIC: new TestPresetEntryV4(async ({ vaults, engine }) => {
    const vault = vaults[0];
    const orig = _.size(await engine.findNotesMeta({ vault }));
    const note1 = NoteUtils.create({
      id: "bar1",
      fname: "bar1",
      created: 1,
      updated: 1,
      vault,
    });
    const note2 = NoteUtils.create({
      id: "bar2",
      fname: "bar2",
      created: 1,
      updated: 1,
      vault,
    });
    const rootNote = (await engine.findNotes({ fname: "root", vault }))[0];
    await engine.bulkWriteNotes({ notes: [note1, note2] });
    const barNote = (await engine.getNoteMeta("bar1")).data!;
    return [
      {
        actual: _.size(await engine.findNotesMeta({ vault })),
        expected: orig + 2,
        msg: "should be 2 more notes",
      },
      {
        expected: barNote.id,
        actual: note1.id,
      },
      {
        expected: barNote.fname,
        actual: note1.fname,
      },
      {
        expected: barNote.vault,
        actual: note1.vault,
      },
      {
        expected: barNote.parent,
        actual: rootNote.id,
      },
    ];
  }),
};
export const ENGINE_BULK_WRITE_NOTES_PRESETS = {
  NOTES,
  SCHEMAS,
};
