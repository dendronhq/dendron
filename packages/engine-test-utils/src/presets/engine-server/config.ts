import { CONSTANTS, ConfigUtils } from "@dendronhq/common-all";
import { DConfig, readYAML, writeYAML } from "@dendronhq/common-server";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import path from "path";

function genDefaultConfig() {
  return ConfigUtils.genDefaultConfig();
}

const WRITE = {
  NEW_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const config = genDefaultConfig();
    ConfigUtils.setPublishProp(config, "copyAssets", false);
    await DConfig.writeConfig({ wsRoot, config });
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    const configOnFile = readYAML(cpath);
    return [
      {
        actual: configOnFile,
        expected: config,
      },
    ];
  }),
};

const GET = {
  DEFAULT_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    writeYAML(cpath, genDefaultConfig());
    const config = DConfig.readConfigSync(wsRoot);
    return [
      {
        actual: config,
        expected: genDefaultConfig(),
      },
    ];
  }),
};

export const ENGINE_CONFIG_PRESETS = {
  WRITE,
  GET,
};
