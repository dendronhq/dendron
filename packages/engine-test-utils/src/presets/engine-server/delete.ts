import {
  CONSTANTS,
  extractNoteChangeEntriesByType,
  NoteChangeEntry,
} from "@dendronhq/common-all";
import { vault2Path } from "@dendronhq/common-server";
import {
  TestPresetEntryV4,
  SCHEMA_PRESETS_V4,
  FileTestUtils,
  EngineTestUtilsV4,
  NoteTestUtilsV4,
  NOTE_PRESETS_V4,
} from "@dendronhq/common-test-utils";
import {
  createCacheEntry,
  DendronEngineClient,
  NotesFileSystemCache,
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { ENGINE_HOOKS, setupBasic } from "./utils";

const SCHEMAS = {
  BASIC: new TestPresetEntryV4(
    async ({ wsRoot, vaults, engine }) => {
      const vault = vaults[0];
      const schemaId = SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.fname;
      const beforeSchema = (await engine.getSchema(schemaId)).data!;
      await engine.deleteSchema(schemaId);
      const afterSchema = (await engine.getSchema(schemaId)).data!;
      return [
        { actual: _.size(beforeSchema.schemas), expected: 2 },
        { actual: afterSchema, expected: undefined },
        {
          actual: await FileTestUtils.assertInVault({
            vault,
            wsRoot,
            nomatch: [`${schemaId}.schema.yml`],
          }),
          expected: true,
        },
      ];
    },
    {
      preSetupHook: setupBasic,
    }
  ),
};
const NOTES = {
  GRANDCHILD_WITH_ALL_STUB_PARENTS: new TestPresetEntryV4(
    async ({ wsRoot, vaults, engine }) => {
      const vault = vaults[0];
      const logger = (engine as DendronEngineClient).logger;
      const cachePath = path.join(
        vault2Path({ wsRoot, vault }),
        CONSTANTS.DENDRON_CACHE_FILE
      );
      const notesCache = new NotesFileSystemCache({ cachePath, logger });
      const keySet = notesCache.getCacheEntryKeys();
      const fooChildNote = (
        await engine.findNotesMeta({ fname: "foo.ch1", vault })
      )[0];
      const resp = await engine.deleteNote(fooChildNote.id);
      const changed = resp.data;
      await engine.init();
      return [
        {
          actual: await EngineTestUtilsV4.checkVault({
            wsRoot,
            vault,
            nomatch: ["foo"],
          }),
          expected: true,
        },
        {
          actual: _.find(
            changed,
            (ent) => ent.status === "delete" && ent.note.fname === "foo.ch1"
          ),
          expected: true,
        },
        {
          actual: _.find(
            changed,
            (ent) => ent.status === "delete" && ent.note.fname === "foo"
          ),
          expected: true,
        },
        {
          actual: _.find(
            changed,
            (ent) => ent.status === "update" && ent.note.fname === "root"
          ),
          expected: true,
        },
        {
          actual: keySet.size,
          expected: 2,
        },
        {
          actual: new NotesFileSystemCache({
            cachePath,
            logger,
          }).getCacheEntryKeys().size,
          expected: 1,
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        await NoteTestUtilsV4.createNote({
          fname: "foo.ch1",
          vault: vaults[0],
          wsRoot,
          body: "",
        });
      },
    }
  ),
  NOTE_NO_CHILDREN: new TestPresetEntryV4(
    async ({ wsRoot, vaults, engine }) => {
      const vault = vaults[0];
      const logger = (engine as DendronEngineClient).logger;
      const notesInVaultBefore = await engine.findNotesMeta({ vault });
      const cachePath = path.join(
        vault2Path({ wsRoot, vault }),
        CONSTANTS.DENDRON_CACHE_FILE
      );
      const notesCache = new NotesFileSystemCache({ cachePath, logger });
      const keySet = notesCache.getCacheEntryKeys();
      const fooChildNote = (
        await engine.findNotesMeta({ fname: "foo.ch1", vault })
      )[0];
      const resp = await engine.deleteNote(fooChildNote.id);

      // Foo's child should be deleted, leaving behind foo and 3 root notes
      const notesInVaultAfter = await engine.findNotesMeta({ vault });
      const fooNote = (await engine.findNotesMeta({ fname: "foo", vault }))[0];
      const changed = resp.data;
      const vpath = vault2Path({ vault, wsRoot });
      await engine.init();
      return [
        { actual: changed![0].note.id, expected: "foo" },
        { actual: _.size(notesInVaultBefore), expected: 3 },
        { actual: _.size(notesInVaultAfter), expected: 2 },
        { actual: fooNote.children, expected: [] },
        {
          actual: _.includes(fs.readdirSync(vpath), "foo.ch1.md"),
          expected: false,
        },
        {
          actual: keySet.size,
          expected: 3,
        },
        {
          actual: new NotesFileSystemCache({
            cachePath,
            logger,
          }).getCacheEntryKeys().size,
          expected: 2,
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        await NoteTestUtilsV4.createNote({
          fname: "foo",
          vault: vaults[0],
          wsRoot,
        });
        await NoteTestUtilsV4.createNote({
          fname: "foo.ch1",
          vault: vaults[0],
          wsRoot,
          body: "",
        });
      },
    }
  ),
  NOTE_WITH_TARGET: new TestPresetEntryV4(
    async ({ vaults, engine }) => {
      const vault = vaults[0];
      const notesInVaultBefore = await engine.findNotesMeta({ vault });
      const betaNoteBefore = await engine.getNoteMeta("beta");
      const betaBackLinksBefore = betaNoteBefore.data!.links.filter(
        (link) => link.type === "backlink"
      );
      await engine.deleteNote("alpha");

      // Alpha be deleted, leaving behind beta and 3 root notes
      const notesInVaultAfter = await engine.findNotesMeta({ vault });
      const betaNoteAfter = await engine.getNoteMeta("beta");
      const betaBackLinksAfter = betaNoteAfter.data!.links.filter(
        (link) => link.type === "backlink"
      );
      return [
        { actual: betaNoteBefore.data?.links.length, expected: 2 },
        { actual: betaBackLinksBefore.length, expected: 1 },
        { actual: _.size(notesInVaultBefore), expected: 3 },
        { actual: _.size(notesInVaultAfter), expected: 2 },
        { actual: betaNoteAfter.data?.links.length, expected: 1 },
        { actual: betaBackLinksAfter.length, expected: 0 },
      ];
    },
    {
      preSetupHook: async ({ wsRoot, vaults }) => {
        await NOTE_PRESETS_V4.NOTE_WITH_TARGET.create({
          wsRoot,
          vault: vaults[0],
        });
        await NOTE_PRESETS_V4.NOTE_WITH_LINK.create({
          wsRoot,
          vault: vaults[0],
        });
      },
    }
  ),
  DOMAIN_CHILDREN: new TestPresetEntryV4(
    async ({ wsRoot, vaults, engine }) => {
      const vault = vaults[0];
      const notesInVaultBefore = await engine.findNotesMeta({ vault });
      const noteToDelete = (
        await engine.findNotesMeta({ fname: "foo", vault })
      )[0];
      const resp = await engine.deleteNote(noteToDelete?.id as string);

      const createEntries = extractNoteChangeEntriesByType(
        resp.data!,
        "create"
      );

      const deleteEntries = extractNoteChangeEntriesByType(
        resp.data!,
        "delete"
      );

      const updateEntries = extractNoteChangeEntriesByType(
        resp.data!,
        "update"
      );

      const notesInVaultAfter = await engine.findNotesMeta({ vault });
      const fooNote = (await engine.findNotesMeta({ fname: "foo", vault }))[0];
      const fooChild = (
        await engine.findNotesMeta({ fname: "foo.ch1", vault })
      )[0];
      const vpath = vault2Path({ vault, wsRoot });

      return [
        {
          actual: updateEntries.length,
          expected: 2,
          msg: "2 updates should happen.",
        },
        {
          actual: deleteEntries.length,
          expected: 1,
          msg: "1 delete should happen.",
        },
        {
          actual: createEntries.length,
          expected: 1,
          msg: "1 create should happen.",
        },
        {
          actual: _.size(notesInVaultBefore),
          expected: 3,
          msg: "Before as root, foo, and foo.ch1",
        },
        {
          actual: _.size(notesInVaultAfter),
          expected: 3,
          msg: "same number of notes as before",
        },
        {
          actual: fooNote.stub,
          expected: true,
          msg: "foo should be a stub",
        },
        {
          actual: fooChild.parent,
          expected: fooNote.id,
          msg: "foo's child should have updated parent id",
        },
        {
          actual: _.includes(fs.readdirSync(vpath), "foo.md"),
          expected: false,
          msg: "note should be deleted",
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        await NoteTestUtilsV4.createNote({
          fname: "foo",
          vault: vaults[0],
          wsRoot,
        });
        await NoteTestUtilsV4.createNote({
          fname: "foo.ch1",
          vault: vaults[0],
          wsRoot,
          body: "",
        });
      },
    }
  ),
  DOMAIN_NO_CHILDREN: new TestPresetEntryV4(
    async ({ wsRoot, vaults, engine }) => {
      const vault = vaults[0];
      const notesInVaultBefore = await engine.findNotesMeta({ vault });
      const logger = (engine as DendronEngineClient).logger;
      const cachePath = path.join(
        vault2Path({ wsRoot, vault }),
        CONSTANTS.DENDRON_CACHE_FILE
      );
      const notesCache = new NotesFileSystemCache({ cachePath, logger });
      const keySet = notesCache.getCacheEntryKeys();
      const noteToDelete = (
        await engine.findNotesMeta({ fname: "foo", vault })
      )[0];
      const resp = await engine.deleteNote(noteToDelete?.id as string);

      const changed = resp.data as NoteChangeEntry[];
      const notesInVaultAfter = await engine.findNotesMeta({ vault });
      const fooNote = (await engine.findNotesMeta({ fname: "foo", vault }))[0];
      const vpath = vault2Path({ vault, wsRoot });
      await engine.init();
      return [
        {
          actual: changed[0].note.fname,
          expected: "root",
          msg: "root updated",
        },
        {
          actual: changed[0].note.children,
          expected: [],
          msg: "root does not have children",
        },
        {
          actual: _.size(notesInVaultBefore),
          expected: 2,
          msg: "Before has root and foo",
        },
        {
          actual: _.size(notesInVaultAfter),
          expected: 1,
          msg: "After has root",
        },
        { actual: fooNote, expected: undefined },
        {
          actual: _.includes(fs.readdirSync(vpath), "foo.md"),
          expected: false,
        },
        {
          actual: keySet.size,
          expected: 2,
        },
        {
          actual: new NotesFileSystemCache({
            cachePath,
            logger,
          }).getCacheEntryKeys().size,
          expected: 1,
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        await NoteTestUtilsV4.createNote({
          fname: "foo",
          vault: vaults[0],
          wsRoot,
        });
      },
    }
  ),
  STALE_CACHE_ENTRY: new TestPresetEntryV4(
    async ({ wsRoot, vaults, engine }) => {
      const vault = vaults[0];
      const logger = (engine as DendronEngineClient).logger;
      const cachePath = path.join(
        vault2Path({ wsRoot, vault }),
        CONSTANTS.DENDRON_CACHE_FILE
      );
      const notesCache = new NotesFileSystemCache({ cachePath, logger });

      // Create random note and write to cache
      const staleNote = await NoteTestUtilsV4.createNote({
        fname: "my-new-note",
        wsRoot,
        vault: vaults[0],
        noWrite: true,
      });
      const cacheEntry = createCacheEntry({
        noteProps: staleNote,
        hash: "123123",
      });
      notesCache.set(staleNote.fname, cacheEntry);
      notesCache.writeToFileSystem();
      const keySet = notesCache.getCacheEntryKeys();

      // Should remove random note from cache
      await engine.init();
      return [
        {
          actual: keySet.size,
          expected: 5,
        },
        {
          actual: keySet.has("my-new-note"),
          expected: true,
        },
        {
          actual: new NotesFileSystemCache({
            cachePath,
            logger,
          }).getCacheEntryKeys().size,
          expected: 4,
        },
        {
          actual: new NotesFileSystemCache({ cachePath, logger })
            .getCacheEntryKeys()
            .has("my-new-note"),
          expected: false,
        },
      ];
    },
    {
      preSetupHook: async (opts) => {
        await ENGINE_HOOKS.setupBasic(opts);
      },
    }
  ),
  MULTIPLE_DELETES: new TestPresetEntryV4(
    async ({ wsRoot, vaults, engine }) => {
      const vault = vaults[0];
      const notesInVaultBefore = await engine.findNotesMeta({ vault });
      const logger = (engine as DendronEngineClient).logger;
      const cachePath = path.join(
        vault2Path({ wsRoot, vault }),
        CONSTANTS.DENDRON_CACHE_FILE
      );
      const notesCache = new NotesFileSystemCache({ cachePath, logger });
      const keySet = notesCache.getCacheEntryKeys();
      const fooChildNote = (
        await engine.findNotesMeta({ fname: "foo.ch1", vault })
      )[0];
      const resp = await engine.deleteNote(fooChildNote.id);
      const changed = resp.data;

      const fooNote = (await engine.findNotesMeta({ fname: "foo", vault }))[0];
      const resp2 = await engine.deleteNote(fooNote.id);
      const changed2 = resp2.data;
      const notesInVaultAfter = await engine.findNotesMeta({ vault });
      await engine.init();
      return [
        {
          actual: _.size(notesInVaultBefore),
          expected: 3,
          msg: "Before has root, foo, and foo.ch1",
        },
        {
          actual: _.size(notesInVaultAfter),
          expected: 1,
          msg: "After has root",
        },
        {
          actual: _.find(
            changed,
            (ent) => ent.status === "delete" && ent.note.fname === "foo.ch1"
          ),
          expected: true,
        },
        {
          actual: _.find(
            changed2,
            (ent) => ent.status === "delete" && ent.note.fname === "foo"
          ),
          expected: true,
        },
        {
          actual: keySet.size,
          expected: 3,
        },
        {
          actual: new NotesFileSystemCache({
            cachePath,
            logger,
          }).getCacheEntryKeys().size,
          expected: 1,
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        await NoteTestUtilsV4.createNote({
          fname: "foo",
          vault: vaults[0],
          wsRoot,
        });
        await NoteTestUtilsV4.createNote({
          fname: "foo.ch1",
          vault: vaults[0],
          wsRoot,
          body: "",
        });
      },
    }
  ),
};
export const ENGINE_DELETE_PRESETS = {
  NOTES,
  SCHEMAS,
};
