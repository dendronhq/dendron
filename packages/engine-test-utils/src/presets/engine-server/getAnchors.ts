import { NoteUtils } from "@dendronhq/common-all";
import {
  TestPresetEntryV4,
  NOTE_PRESETS_V4,
} from "@dendronhq/common-test-utils";
import _ from "lodash";

const SCHEMAS = {};
const NOTES = {
  /**
   * Check that we can get both roots from both vaults
   */
  REGULAR: new TestPresetEntryV4(
    async ({ vaults, engine }) => {
      const note = NoteUtils.getNoteByFnameFromEngine({
        fname: NOTE_PRESETS_V4.NOTE_WITH_ANCHOR_TARGET.fname,
        engine,
        vault: vaults[0],
      });

      const { data } = await engine.getAnchors({
        note: note!,
      });
      return [
        {
          actual: data,
          expected: {
            "^8a": {
              column: 5,
              line: 8,
              type: "block",
              value: "8a",
            },
            h1: {
              column: 0,
              line: 7,
              depth: 1,
              text: "H1",
              type: "header",
              value: "h1",
            },
            h2: {
              column: 0,
              line: 8,
              depth: 1,
              text: "H2",
              type: "header",
              value: "h2",
            },
            h3: {
              column: 0,
              line: 9,
              depth: 1,
              text: "H3",
              type: "header",
              value: "h3",
            },
          },
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        await NOTE_PRESETS_V4.NOTE_WITH_ANCHOR_TARGET.create({
          wsRoot,
          vault: vaults[0],
        });
      },
    }
  ),
};
export const ENGINE_GET_ANCHORS_PRESETS = {
  NOTES,
  SCHEMAS,
};
