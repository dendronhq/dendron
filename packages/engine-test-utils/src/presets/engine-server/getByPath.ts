import { NoteUtils } from "@dendronhq/common-all";
import {
  TestPresetEntryV4,
  NOTE_PRESETS_V4,
} from "@dendronhq/common-test-utils";

const SCHEMAS = {};
const NOTES = {
  /**
   * Check that we can get both roots from both vaults
   */
  ROOT: new TestPresetEntryV4(async ({ vaults, engine, wsRoot }) => {
    const vault = vaults[0];
    const vault2 = vaults[1];
    const root = NoteUtils.getNoteByFnameV5({
      fname: "root",
      notes: engine.notes,
      vault,
      wsRoot,
    });
    const root2 = NoteUtils.getNoteByFnameV5({
      fname: "root",
      notes: engine.notes,
      vault: vault2,
      wsRoot,
    });
    const { data } = await engine.getNoteByPath({ npath: "root", vault });
    const { data: data2 } = await engine.getNoteByPath({
      npath: "root",
      vault: vault2,
    });
    return [
      {
        actual: data?.changed,
        expected: [],
      },
      {
        actual: data?.note,
        expected: root,
      },
      {
        actual: data2?.note,
        expected: root2,
      },
    ];
  }),
  EXISTING_NOTE: new TestPresetEntryV4(
    async ({ vaults, engine }) => {
      const vault = vaults[0];
      const note = NoteUtils.getNoteByFnameV5({
        fname: "foo",
        notes: engine.notes,
        vault,
        wsRoot: engine.wsRoot,
      });
      const { data } = await engine.getNoteByPath({ npath: "foo", vault });
      return [
        {
          actual: data?.changed,
          expected: [],
        },
        {
          actual: data?.note,
          expected: note,
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        const vault = vaults[0];
        await NOTE_PRESETS_V4.NOTE_SIMPLE.create({ vault, wsRoot });
      },
    }
  ),
  NOTE_WITH_CAPS_AND_SPACES: new TestPresetEntryV4(
    async ({ vaults, engine }) => {
      const vault = vaults[0];
      const note = NoteUtils.getNoteByFnameV5({
        fname: "000 Index",
        notes: engine.notes,
        vault,
        wsRoot: engine.wsRoot,
      });
      const { data } = await engine.getNoteByPath({
        npath: "000 Index",
        vault,
      });
      return [
        {
          actual: data?.changed,
          expected: [],
        },
        {
          actual: data?.note,
          expected: note,
        },
      ];
    },
    {
      preSetupHook: async ({ vaults, wsRoot }) => {
        const vault = vaults[0];
        await NOTE_PRESETS_V4.NOTE_WITH_CAPS_AND_SPACE.create({
          vault,
          wsRoot,
        });
      },
    }
  ),
  NEW_NOTE: new TestPresetEntryV4(async ({ vaults, engine }) => {
    const vault = vaults[0];
    const { data } = await engine.getNoteByPath({
      npath: "bar",
      vault,
      createIfNew: true,
    });
    return [
      {
        actual: data?.changed.map((ent) => ({
          fname: ent.note.fname,
          status: ent.status,
        })),
        expected: [
          { fname: "root", status: "update" },
          { fname: "bar", status: "create" },
        ],
      },
      {
        actual: data?.note?.fname,
        expected: "bar",
      },
    ];
  }),
  NEW_NOTE_WITH_OVERRIDES: new TestPresetEntryV4(async ({ vaults, engine }) => {
    const vault = vaults[0];
    const { data } = await engine.getNoteByPath({
      npath: "bar",
      vault,
      createIfNew: true,
      overrides: {
        title: "bar title",
      },
    });
    return [
      {
        actual: data?.changed.map((ent) => ({
          fname: ent.note.fname,
          status: ent.status,
        })),
        expected: [
          { fname: "root", status: "update" },
          { fname: "bar", status: "create" },
        ],
      },
      {
        actual: data?.note?.fname,
        expected: "bar",
      },
      {
        actual: data?.note?.title,
        expected: "bar title",
      },
    ];
  }),
};
export const ENGINE_GET_NOTE_BY_PATH_PRESETS = {
  NOTES,
  SCHEMAS,
};
