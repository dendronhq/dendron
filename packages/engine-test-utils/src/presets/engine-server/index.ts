import { ENGINE_CONFIG_PRESETS } from "./config";
import { ENGINE_DELETE_PRESETS } from "./delete";
import { ENGINE_INFO_PRESETS } from "./info";
import { ENGINE_INIT_PRESETS } from "./init";
import { ENGINE_GET_NOTE_BLOCKS_PRESETS } from "./getNoteBlocks";
import NOTE_REF from "./note-refs";
import { ENGINE_QUERY_PRESETS } from "./query";
import { ENGINE_RENAME_PRESETS } from "./rename";
import { ENGINE_BULK_WRITE_NOTES_PRESETS } from "./bulkWriteNotes";
import { ENGINE_RENDER_PRESETS } from "./render";
import { ENGINE_WRITE_PRESETS, ENGINE_WRITE_PRESETS_MULTI } from "./write";
import _ from "lodash";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";

export { ENGINE_HOOKS, ENGINE_HOOKS_BASE, ENGINE_HOOKS_MULTI } from "./utils";
export { ENGINE_RENAME_PRESETS };
export { ENGINE_QUERY_PRESETS };
export { ENGINE_WRITE_PRESETS };
export { ENGINE_CONFIG_PRESETS };

export const ENGINE_SERVER = {
  NOTE_REF,
  ENGINE_WRITE_PRESETS,
  ENGINE_INIT_PRESETS,
  ENGINE_DELETE_PRESETS,
  ENGINE_INFO_PRESETS,
  ENGINE_RENAME_PRESETS,
  ENGINE_GET_NOTE_BLOCKS_PRESETS,
  ENGINE_QUERY_PRESETS,
  ENGINE_BULK_WRITE_NOTES_PRESETS,
  ENGINE_RENDER_PRESETS,
};

type TestPresetEntry = TestPresetEntryV4;
type TestPresetDict = { [key: string]: TestPresetEntry };

/**
 * 
 @example
 *  test("", async () => {
 *    const TestCase= getPreset({presets: ENGINE_PRESETS, nodeType, presetName: "init", key: 'BAD_SCHEMA'})
 *    const { testFunc, ...opts } = TestCase;;
 *    await runEngineTestV5(testFunc, { ...opts, expect });
 *});
 * @param param0 
 * @returns 
 */
export const getPreset = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetMulti = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS_MULTI;
  presetName: string;
  nodeType: "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetGroup = ({
  presets,
  presetName,
  nodeType,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  return ent.presets[nodeType] as TestPresetDict;
};

// ^iygzn9r2758w
export const ENGINE_PRESETS = [
  { name: "write", presets: ENGINE_SERVER.ENGINE_WRITE_PRESETS },
  {
    name: "bulkWriteNotes",
    presets: ENGINE_SERVER.ENGINE_BULK_WRITE_NOTES_PRESETS,
  },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
  { name: "info", presets: ENGINE_SERVER.ENGINE_INFO_PRESETS },
  {
    name: "getNoteBlocks",
    presets: ENGINE_SERVER.ENGINE_GET_NOTE_BLOCKS_PRESETS,
  },
  { name: "render", presets: ENGINE_SERVER.ENGINE_RENDER_PRESETS },
];

export const ENGINE_PRESETS_MULTI = [
  { name: "write", presets: ENGINE_WRITE_PRESETS_MULTI },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
];
