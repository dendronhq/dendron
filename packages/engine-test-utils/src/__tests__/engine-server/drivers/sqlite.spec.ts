import { NoteUtils } from "@dendronhq/common-all";
import { SQLiteMetadataStore } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import sinon from "sinon";
import { ENGINE_HOOKS, runEngineTestV5 } from "../../..";
// import os from "os";

// current issue with windows test
// pathname is generated in the following format
// unable to open database file: D:///c%3A/Users/RUNNER~1/AppData/Local/Temp/tmp-6392-F8nMLdBlS5Ty/metadata.db\n
// const describeSkipWindows =
//   os.platform() === "win32" ? describe.skip : describe;

describe.skip("GIVEN sqlite store", () => {
  afterEach(async () => {
    await SQLiteMetadataStore.prisma().$disconnect();
  });
  jest.setTimeout(10e3);

  test("WHEN initialize, THEN metadata has all notes", async () => {
    await runEngineTestV5(
      async ({ wsRoot, engine }) => {
        const dirList = fs.readdirSync(wsRoot);
        expect(dirList).toMatchSnapshot();
        expect(dirList.includes("metadata.db")).toBeTruthy();
        const notes = await SQLiteMetadataStore.prisma().note.findMany();
        const engineNotes = await engine.findNotesMeta({ excludeStub: false });
        expect(engineNotes.length).toEqual(notes.length);
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
        modConfigCb: (config) => {
          config.workspace.metadataStore = "sqlite";
          return config;
        },
      }
    );
  });

  test("WHEN initialize mult times, THEN metadata has all notes", async () => {
    await runEngineTestV5(
      async ({ wsRoot, engine }) => {
        const dirList = fs.readdirSync(wsRoot);
        expect(dirList).toMatchSnapshot();
        expect(dirList.includes("metadata.db")).toBeTruthy();
        const notes = await SQLiteMetadataStore.prisma().note.findMany();
        const engineNotes = await engine.findNotesMeta({ excludeStub: false });
        expect(engineNotes.length).toEqual(notes.length);

        const { error } = await engine.init();
        expect(error).toBeFalsy();
        const createAllSpy = sinon.spy(SQLiteMetadataStore, "createAllTables");
        expect(createAllSpy.calledOnce).toBeFalsy();
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
        modConfigCb: (config) => {
          config.workspace.metadataStore = "sqlite";
          return config;
        },
      }
    );
  });

  test("WHEN initialize mult times and notes have changed, THEN metadata is updated", async () => {
    await runEngineTestV5(
      async ({ wsRoot, engine }) => {
        const dirList = fs.readdirSync(wsRoot);
        expect(dirList).toMatchSnapshot();
        expect(dirList.includes("metadata.db")).toBeTruthy();
        const notes = await SQLiteMetadataStore.prisma().note.findMany();
        const engineNotes = await engine.findNotesMeta({ excludeStub: false });
        expect(engineNotes.length).toEqual(notes.length);
        const newNote = NoteUtils.create({
          id: "new-note",
          fname: "new-note",
          vault: engine.vaults[0],
        });
        await engine.writeNote(newNote);

        const { error } = await engine.init();
        expect(error).toBeFalsy();
        const foundNote = await SQLiteMetadataStore.prisma().note.findFirst({
          where: { fname: "new-note" },
        });
        expect(foundNote?.fname).toEqual(newNote.fname);
      },
      {
        expect,
        preSetupHook: ENGINE_HOOKS.setupBasic,
        modConfigCb: (config) => {
          config.workspace.metadataStore = "sqlite";
          return config;
        },
      }
    );
  });
});
