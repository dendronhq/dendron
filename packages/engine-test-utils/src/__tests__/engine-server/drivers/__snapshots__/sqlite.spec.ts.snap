// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN sqlite store WHEN initialize mult times and notes have changed, THEN metadata is updated 1`] = `
Array [
  "dendron.code-workspace",
  "dendron.yml",
  "hooks",
  "metadata.db",
  "vault1",
  "vault2",
  "vault3",
]
`;

exports[`GIVEN sqlite store WHEN initialize mult times, THEN metadata has all notes 1`] = `
Array [
  "dendron.code-workspace",
  "dendron.yml",
  "hooks",
  "metadata.db",
  "vault1",
  "vault2",
  "vault3",
]
`;

exports[`GIVEN sqlite store WHEN initialize, THEN metadata has all notes 1`] = `
Array [
  "dendron.code-workspace",
  "dendron.yml",
  "hooks",
  "metadata.db",
  "vault1",
  "vault2",
  "vault3",
]
`;
